temporalio/worker/workflow_sandbox/restrictions.py

```diff
@@ -0,0 +1,971 @@
+        if instance is None:
+            if self.class_value is not None:
+                return self.class_value
+
+            return self
+
+        try:
+            state: _RestrictionState = object.__getattribute__(
+                instance, "__temporal_state"
+            )
+        except RuntimeError:
+            if self.fallback_func is None:
+                raise
+
+            fallback = self.fallback_func.__get__(instance, owner)
+
+            if self.is_attr:
+                # __class__ and __doc__ are attributes, not methods.
+                # Call the fallback to get the value.
+                return fallback()
+
+            return fallback
+
+        if self.bind_func is not None:
+            return self.bind_func(instance, state.obj)
+
+        return getattr(state.obj, self.name)
+
+    def __repr__(self) -> str:
+        return f"proxy {self.name}"
+
+    def __call__(self, instance: _RestrictedProxy, *args: Any, **kwargs: Any) -> Any:
+        """Support calling unbound methods from the class. For example,
+        this happens with ``copy.copy``, which does
+        ``type(x).__copy__(x)``. ``type(x)`` can't be proxied, so it
+        returns the proxy type and descriptor.
+        """
+        return self.__get__(instance, type(instance))(*args, **kwargs)
+
+
+class _RestrictedProxyIOp(_RestrictedProxyLookup):
+    __slots__ = ()
+
+    def __init__(
+        self,
+        access_func: Optional[Callable] = None,
+        *,
+        fallback_func: Optional[Callable] = None,
+    ) -> None:
+        super().__init__(access_func, fallback_func=fallback_func)
+
+        def bind_f(instance: _RestrictedProxy, obj: Any) -> Callable:
+            def i_op(self: Any, other: Any) -> _RestrictedProxy:
+                f(self, other)  # type: ignore
+                return instance
+
+            return i_op.__get__(obj, type(obj))  # type: ignore
+
+        self.bind_f = bind_f
+
+
+_OpF = TypeVar("_OpF", bound=Callable[..., Any])
+
+
+def _l_to_r_op(op: _OpF) -> _OpF:
+    """Swap the argument order to turn an l-op into an r-op."""
+
+    def r_op(obj: Any, other: Any) -> Any:
+        return op(other, obj)
+
+    return cast(_OpF, r_op)
+
+
+def _is_restrictable(v: Any) -> bool:
+    return v is not None and not isinstance(
+        v, (bool, int, float, complex, str, bytes, bytearray)
+    )
+
+
+class _RestrictedProxy:
+    def __init__(
+        self, name: str, obj: Any, context: RestrictionContext, matcher: SandboxMatcher
+    ) -> None:
+        _trace("__init__ on %s", name)
+        _RestrictionState(
+            name=name, obj=obj, context=context, matcher=matcher
+        ).set_on_proxy(self)
+
+    def __getattribute__(self, __name: str) -> Any:
+        state = _RestrictionState.from_proxy(self)
+        _trace("__getattribute__ %s on %s", __name, state.name)
+        state.assert_child_not_restricted(__name)
+        ret = object.__getattribute__(self, "__getattr__")(__name)
```

Comment:
This might be the highest density of underscores I've ever seen in a code block lol

Reply:
It's definitely confusing. It's how you prevent accidental recursion when accessing `self` after you have overridden `__getattribute__` (ref https://docs.python.org/3/reference/datamodel.html#object.__getattribute__).